app/src/rgb_underglow.c

```diff
@@ -292,101 +265,144 @@ int zmk_rgb_underglow_save_state() {
+#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_EXT_POWER)
+    if (ext_power != NULL) {
+        int rc = ext_power_enable(ext_power);
+        if (rc != 0) {
+            LOG_ERR("Unable to enable EXT_POWER: %d", rc);
+        }
     }
+#endif
 
+    state.on = true;
     state.animation_step = 0;
+    k_timer_start(&underglow_tick, K_NO_WAIT, K_MSEC(50));
 
     return zmk_rgb_underglow_save_state();
 }
 
-int zmk_rgb_underglow_toggle() {
+int zmk_rgb_underglow_off() {
     if (!led_strip)
         return -ENODEV;
 
-    state.on = !state.on;
-
 #if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_EXT_POWER)
     if (ext_power != NULL) {
-        int rc;
-
-        if (state.on) {
-            rc = ext_power_enable(ext_power);
-        } else {
-            rc = ext_power_disable(ext_power);
-        }
-
+        int rc = ext_power_disable(ext_power);
         if (rc != 0) {
-            LOG_ERR("Unable to toggle EXT_POWER: %d", rc);
+            LOG_ERR("Unable to disable EXT_POWER: %d", rc);
         }
     }
 #endif
 
-    if (state.on) {
-        state.animation_step = 0;
-        k_timer_start(&underglow_tick, K_NO_WAIT, K_MSEC(50));
-    } else {
-        zmk_rgb_underglow_off();
+    for (int i = 0; i < STRIP_NUM_PIXELS; i++) {
+        pixels[i] = (struct led_rgb){r : 0, g : 0, b : 0};
+    }
+
+    led_strip_update_rgb(led_strip, pixels, STRIP_NUM_PIXELS);
+
+    k_timer_stop(&underglow_tick);
+    state.on = false;
 
-        k_timer_stop(&underglow_tick);
+    return zmk_rgb_underglow_save_state();
+}
+
+int zmk_rgb_underglow_cycle_effect(int direction) {
+    if (!led_strip)
+        return -ENODEV;
+
+    if (state.current_effect == 0 && direction < 0) {
+        state.current_effect = UNDERGLOW_EFFECT_NUMBER - 1;
+        return 0;
     }
 
+    state.current_effect += direction;
+
+    if (state.current_effect >= UNDERGLOW_EFFECT_NUMBER) {
+        state.current_effect = 0;
+    }
+
+    state.animation_step = 0;
+
     return zmk_rgb_underglow_save_state();
 }
 
-int zmk_rgb_underglow_set_hsb(uint16_t hue, uint8_t saturation, uint8_t brightness) {
-    if (hue > 360 || saturation > 100 || brightness > 100) {
+int zmk_rgb_underglow_toggle() {
+    return state.on ? zmk_rgb_underglow_off() : zmk_rgb_underglow_on();
+}
+
+int zmk_rgb_underglow_set_hsb(struct zmk_led_hsb color) {
+    if (color.h > HUE_MAX || color.s > SAT_MAX || color.b > BRT_MAX) {
         return -ENOTSUP;
     }
 
-    state.hue = hue;
-    state.saturation = saturation;
-    state.brightness = brightness;
+    state.color = color;
 
     return 0;
 }
 
-int zmk_rgb_underglow_change_hue(int direction) {
-    if (!led_strip)
-        return -ENODEV;
+struct zmk_led_hsb zmk_rgb_underglow_calc_hue(int direction) {
+    struct zmk_led_hsb color = state.color;
 
-    if (state.hue == 0 && direction < 0) {
-        state.hue = 360 - CONFIG_ZMK_RGB_UNDERGLOW_HUE_STEP;
-        return 0;
+    color.h += HUE_MAX + (direction * CONFIG_ZMK_RGB_UNDERGLOW_HUE_STEP);
+    color.h %= HUE_MAX;
```

Comment:
```suggestion
    color.h %= HUE_MAX;
    
    if (color.h < 0) {
        color.h += HUE_MAX;
    }
```
I think this is necessary

Reply:
@Nicell  Double check the line above this one. It adds an *extra* `HUE_MAX` before performing the modulo.

Reply:
Ah! Missed that. Clever, makes sense.